### PR TITLE
Extend ScheduleActivityTask decision attributes with a requestLocalDispatch field

### PR DESCRIPTION
**Pull request by mkolodezny**
This is basically the prerequisite for the activity dispatch optimization feature to allow worker to dispatch activity tasks through local tunnel after decisions are made. This is a performance optimization to skip activity scheduling efforts.

The follow up is to populate ActivityLocalDispatchInfo response struct for each activity such as: ScheduledTimestamp, StartedTimestamp and probably StartedEventId if needed.     

The corresponding IDL pr https://github.com/uber/cadence-idl/pull/45

**How did you test it?**
unit tests
e2e testing

**Potential risks**
No risk. It is off by default

### CODE REVIEW

File: common/service/dynamicconfig/constants.go [hunk 1]
```diff
@@ -280,6 +280,7 @@ var keys = map[Key]string{
 	NotifyFailoverMarkerInterval:                          "history.NotifyFailoverMarkerInterval",
 	NotifyFailoverMarkerTimerJitterCoefficient:            "history.NotifyFailoverMarkerTimerJitterCoefficient",
 	EnableDropStuckTaskByDomainID:                         "history.DropStuckTaskByDomain",
+	EnableRequestActivityLocalDispatchByDomain:            "history.enableRequestActivityLocalDispatchByDomain",
```
Comment (meiliang86) [inline common/service/dynamicconfig/constants.go:283]: EnableRequestActivityLocalDispatchByDomain -> EnableActivityLocalDispatchByDomain

File: common/service/dynamicconfig/constants.go [hunk 2]
```diff
@@ -904,6 +905,9 @@ const (
 
 	// lastKeyForTest must be the last one in this const group for testing purpose
 	lastKeyForTest
+
+	// EnableRequestActivityLocalDispatchByDomain indicates if client can request to start activity locally
+	EnableRequestActivityLocalDispatchByDomain
```
Comment (meiliang86) [inline common/service/dynamicconfig/constants.go:910]: Same name change here

File: common/service/dynamicconfig/constants.go [hunk 3]
```diff
@@ -904,6 +905,9 @@ const (
 
 	// lastKeyForTest must be the last one in this const group for testing purpose
 	lastKeyForTest
+
+	// EnableRequestActivityLocalDispatchByDomain indicates if client can request to start activity locally
```
Comment (meiliang86) [inline common/service/dynamicconfig/constants.go:909]: // EnableActivityLocalDispatchByDomain allows worker to dispatch activity tasks through local tunnel after decisions are made. This is an performance optimization to skip activity scheduling efforts.

File: service/history/config/config.go [hunk 4]
```diff
@@ -254,6 +254,9 @@ type Config struct {
 	NotifyFailoverMarkerInterval               dynamicconfig.DurationPropertyFn
 	NotifyFailoverMarkerTimerJitterCoefficient dynamicconfig.FloatPropertyFn
 	EnableGracefulFailover                     dynamicconfig.BoolPropertyFn
+
+	// Activity dispatch optimization
```
Comment (meiliang86) [inline service/history/config/config.go:258]: Apply same comment and name change here.

Comment (meiliang86) [reply service/history/config/config.go:258]: Same below.

File: service/history/decisionTaskHandler.go [hunk 5]
```diff
@@ -219,9 +220,17 @@ func (handler *decisionTaskHandlerImpl) handleDecisionScheduleActivity(
 		return err
 	}
 
-	_, _, err = handler.mutableState.AddActivityTaskScheduledEvent(handler.decisionTaskCompletedID, attr)
+	dispatchLocally := handler.config.EnableRequestActivityLocalDispatchByDomain(handler.domainEntry.GetInfo().Name) &&
```
Comment (meiliang86) [inline service/history/decisionTaskHandler.go:223]: This logic should be put inside AddActivityTaskScheduledEvent

Comment (meiliang86) [reply service/history/decisionTaskHandler.go:223]: You have TODO below to return additional information anyways. So you can use that to decide if you need to ad the started event.

**Comment by meiliang86** [comment]
Any additional unit test you can add?

File: service/history/decisionTaskHandler.go (hunk 5 above)
Comment (mkolodezny) [reply service/history/decisionTaskHandler.go]: agree, lets to update decision handler func's to return interface{} to collect all that info

File: .gen/proto/gogo.pb.go [hunk 6]
```diff
@@ -1,1321 +0,0 @@
-// The MIT License (MIT)
```
Comment (yux0) [inline .gen/proto/gogo.pb.go:1]: why we remove this one?

File: .gen/proto/message.pb.go [hunk 7]
```diff
@@ -1,15581 +0,0 @@
-// The MIT License (MIT)
```
Comment (yux0) [inline .gen/proto/message.pb.go]: same?

File: service/history/execution/history_builder_test.go [hunk 8]
```diff
@@ -841,7 +841,8 @@ func (s *historyBuilderSuite) addActivityTaskScheduledEvent(decisionCompletedID
 			HeartbeatTimeoutSeconds:       common.Int32Ptr(hearbeatTimeout),
 			StartToCloseTimeoutSeconds:    common.Int32Ptr(1),
 			RetryPolicy:                   retryPolicy,
-		})
+		},
+		false)
```
Comment (yux0) [inline service/history/execution/history_builder_test.go:845]: Add some test with true case?

**Review by yycptt — COMMENTED** [review]
Looks good to me! I think it may worth to add an integration test for the e2e behavior so that it won't be break accidentally. I am not sure how much change is needed for the integration test framework though.

**Comment by mkolodezny** [comment]
> Looks good to me! I think it may worth to add an integration test for the e2e behavior so that it won't be break accidentally. I am not sure how much change is needed for the integration test framework though.

sure, thats scheduled as a follow up on client side.

**Comment by mkolodezny** [comment]
> Any additional unit test you can add?

added

**Review by meiliang86 — APPROVED** [review]

**Pull request merged**